### PR TITLE
task 7, 8, 9, 10, 11, 12, 13 done

### PR DESCRIPTION
**Pull request by golovko**

### CODE REVIEW

File: controllers/comments.controller.js [hunk 1]
```diff
@@ -17,4 +21,26 @@ exports.getComments = (req, res, next) => {
     .catch((err)=>{
         next(err);
     })
+}
+
+exports.postComments = (req, res, next) => {
+    const newComment = req.body;
+    if(!newComment.body || !newComment.username) {
+        next({errorCode: 400, errorMessage: 'Bad request. No username or body properties'})
```
Comment (aloush-dev) [inline controllers/comments.controller.js]: we shouldn't really be invoking next here.
We should be using Promise.reject and our error handlers should invoke next until the correct error handler is hit.

File: controllers/comments.controller.js [hunk 2]
```diff
@@ -17,4 +21,26 @@ exports.getComments = (req, res, next) => {
...
+    }
+    newComment.article_id = Number(req.params.article_id);
+    const promises = [
+        checkExists('articles', 'article_id', newComment.article_id ),
+        checkExists('users', 'username', newComment.username ) ]
```
Comment (aloush-dev) [inline controllers/comments.controller.js]: I really like you've made a reuseable function for checkExists.
But we shouldn't need to use it for users as PSQL will throw us an error if the user doesn't exist in the database.

File: models/comments.model.js [hunk 3]
```diff
@@ -10,24 +11,22 @@ exports.fetchComments = (articleId) => {
...
+        author: newComment.username, 
+        body: newComment.body, 
+        votes: 0,
+        article_id: newComment.article_id
+    }
+    const commentValues = Object.values(comment);
     return db.query(`
-        SELECT * FROM articles WHERE article_id = $1
-    `, [articleId])
+        INSERT INTO comments(author, body, votes, article_id)
```
Comment (aloush-dev) [inline models/comments.model.js]: not necessary to send votes as our DB will create the votes property with a default value of 0

File: controllers/articles.controller.js [hunk 4]
```diff
@@ -19,4 +19,22 @@ exports.getArticles = (req, res, next) => {
     .catch((err)=>{
         next(err);
     })
+}
+
+// patch
+exports.patchArticle = (req, res, next) => {
+    if(isNaN(req.body.inc_votes)) {
+        next({errorCode: 400, errorMessage: 'Invalid body'})
+    }
```
Comment (aloush-dev) [inline controllers/articles.controller.js]: PSQL should throw an error here you could use

File: models/articles.model.js [hunk 5]
```diff
@@ -40,4 +41,20 @@ exports.fetchArticles = () => {
         }
         return data.rows;
     })
+}
+
+exports.updateArticle = (article) => {
+    return fetchArticleById(article.article_id)
```
Comment (aloush-dev) [inline models/articles.model.js]: I assume you're using this to test if the article exists which is a good way to go about it.
But have you not previously made a check exists function?
Either way is good, but be consistent when choosing which way to go with it.

File: models/articles.model.js [hunk 6]
```diff
@@ -40,4 +41,20 @@ exports.fetchArticles = () => {
...
+    .then((data) => {
+        return db.query(`
+        UPDATE articles
+        SET votes = $1
+        WHERE article_id = $2
+        RETURNING *;
+    `, [data.votes + article.inc_votes, article.article_id])
```
Comment (aloush-dev) [inline models/articles.model.js]: Ah i see you're doing this to get the current votes.
You can use SQL and do something like SET votes = votes + $1

File: controllers/comments.controller.js [hunk 7]
```diff
@@ -43,4 +43,14 @@ exports.postComments = (req, res, next) => {
     .catch((err)=>{
         next(err);
     })
+}
+
+exports.deleteComment = (req, res, next) => {
+    return removeComment(req.params.comment_id)
+    .then(() =>{
+        res.status(204).send({message: 'Deleted'});
```
Comment (aloush-dev) [inline controllers/comments.controller.js]: Sending a status back would be enough here.
Anything like a message would be better done in the front end.

File: models/users.model.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+exports.fetchUsers = () => {
+    return db.query(`
+    SELECT *
+    FROM users 
+    `)
+    .then((data) => {
+        if(data.rows.length === 0) {
+            return Promise.reject({errorCode: 404, errorMessage: 'User not found'})
+        }
+        return data.rows;
```
Comment (aloush-dev) [inline models/users.model.js]: This model is for returning all users.
This error message would make me think I am trying to request a specific user.

File: models/articles.model.js [hunk 9]
```diff
@@ -26,6 +32,7 @@ exports.fetchArticles = () => {
     CAST(COUNT(comments.comment_id) AS INTEGER) AS comment_count 
     FROM comments 
     JOIN articles ON articles.article_id = comments.article_id
+    ${whereClause}
```
Comment (aloush-dev) [inline models/articles.model.js:53]: You don't want to be using template literals in out query string.
https://notes.northcoders.com/courses/js-back-end/complex-queries#optional-where-clauses

These notes will help you build this up in a better format.

**Review by aloush-dev — CHANGES_REQUESTED** [review]
A lot to go through here since you've sent so much through for review please see and address comments.
Testing is really good throughout.

Topic query needs refactoring to not include tmeplate literals.

Your handling of erros is not very consistent, if we are 'making' custom errors, it should usually be done from a model and it should be using a Promise.reject which will get picked up by the controllers catch block and invoke next. You shouldn't directly invoke next.

Please see comments throughout and update as needed.

File: models/articles.model.js (hunk 9 above)
Comment (golovko) [reply models/articles.model.js:53]: i use  whereClause = `WHERE articles.topic = $1`; to insert safely the value from url/request body  and then using literal just to dynamically insert whole clause in query. Is it anyway unsafe to use literals like this?

File: controllers/comments.controller.js (hunk 1 above)
Comment (golovko) [reply controllers/comments.controller.js]: fixed

File: controllers/comments.controller.js (hunk 2 above)
Comment (golovko) [reply controllers/comments.controller.js]: fixed

File: models/comments.model.js (hunk 3 above)
Comment (golovko) [reply models/comments.model.js]: fixed

File: controllers/articles.controller.js (hunk 4 above)
Comment (golovko) [reply controllers/articles.controller.js]: i moved this check to model

File: models/articles.model.js (hunk 6 above)
Comment (golovko) [reply models/articles.model.js]: refactored

File: models/users.model.js (hunk 8 above)
Comment (golovko) [reply models/users.model.js]: typo =))
userS not found

**Review by robert112233 — APPROVED** [review]
I think it's safe how you have solved it, the literal is already sanitized so it's safe to use 👍

File: __tests__/endpoints.test.js [hunk 10]
```diff
@@ -72,6 +72,16 @@ describe('GET /api/articles', () => {
...
   });
+  test('should return article with a comment_count field', () => { 
+    return query(app)
+      .get('/api/articles/1')
+      .expect(200)
+      .then((res) => {
+          const article = res.body;
+          expect(article.comment_count).toBe(11);
+      });
+  });
```
Comment (robert112233) [inline __tests__/endpoints.test.js:83]: Nicely done

**Review by robert112233 — APPROVED** [review]
Good to merge

**Review by robert112233 — APPROVED** [review]
Readme looks great, happy to merge

**Review by robert112233 — APPROVED** [review]
Your hosted site works great! Well done

**Pull request merged**